src/Tizen.NUI.Scene3D/src/public/Controls/SceneView.cs

```diff
@@ -573,6 +622,27 @@ public void ResetResolution()
             if (NDalicPINVOKE.SWIGPendingException.Pending) throw NDalicPINVOKE.SWIGPendingException.Retrieve();
         }
 
+        /// <summary>
+        /// Requests to capture this SceneView with the Camera.
+        /// </summary>
+        /// <param name="camera">Camera to be used for capture.</param>
+        /// <param name="size">captured size.</param>
+        /// <remarks>
+        /// The input camera should not be used for any other purpose during Capture.
+        /// (Simultaneous usage elsewhere may result in incorrect rendering.)
+        /// The camera is required to be added in this SceneView. (Not need to be a selected camera)
+        /// If the SceneView is disconnected from Scene, the left capture requests are canceled.
+        /// </remarks>
+        /// <returns> capture id that id unique value to distinguish each requiest.</returns>
+        // This will be public opened after ACR done. (Before ACR, need to be hidden as Inhouse API)
+        [EditorBrowsable(EditorBrowsableState.Never)]
+        public int Capture(Scene3D.Camera camera, Vector2 size)
```

Comment:
Dali 구조와 상관 없이 순수 API사용자 관점에서 Capture API의 위치는 camera객체가 갖고 있어야 하는게 아닐까요?

Reply:
특정 행위에 대한 결과물이 비동기적으로 전달되는 경우에는 
C#에서는 이벤트보다는 Task를 사용하는 것이 좋기는 합니다. 
```suggestion
        public Task<ImageUrl> Capture(Scene3D.Camera camera, Vector2 size)
```

이런 API가 제공되면 앱 개발자는 이렇게 단순하게 코드 작성이 가능해 집니다.
``` c#
 var url = await sceneView.Caputre(camea, new Vector2(100, 100));
 imageView.ResourceUrl = url.ToString();
```

Reply:
ImageUrl class 는 UI Thread 에서만 생성되어야하는 class 이기 때문에 Task 로 만들 수 없습니다. ImageUrl class 는 외부 요인에 의해 생성된 이미지요소(ex : Texture, NativeImage 등,)의 lifecycle을 관리하기위한 class 이기 때문입니다.

Reply:
Scene3D 개발자 관점에서, 카메라를 가져오고 설정하는 API 는 모두 SceneView에 있는데 Capture 만 Camera API 로 빠져있는 것이 더 어색해 보입니다.

하나의 Camera 가 동시에 여러 SceneView 에 속해있을 수도 없고, 또한 Camera 자기 자신은 자기가 속해있는 SceneView 도 모르는 상황으로, 정말 Camera 관련 속성들만 제어/관리 하고 있습니다. 이런 상태에서 SceneView 에 의존성이 생기는 API 하나를 추가하는 것은 별로 좋은 구조로 보이지 않습니다.

View.SetRenderEffect(BackgroundBlurEffect) 처럼 RenderEffect 에 대한 구현을 Vjew 가 아니라 RenderEffect 에서 설정하는 것처럼 느껴집니다.

Reply:
> Dali 구조와 상관 없이 순수 API사용자 관점에서 Capture API의 위치는 camera객체가 갖고 있어야 하는게 아닐까요?

말씀하신 형태도 API관점에서 좋은 방법이라고 생각합니다만, 몇가지 문제가 있는데요.

예를들어, Camera가 화면을 촬영하기 위해서 필요한 몇가지 속성 중에 종횡비가 있습니다.
다른 3D 엔진의 경우에는 Window종횡비가 곧 3D Scene의 종횡비이기 때문에 Camera는 대부분의 경우 Window 데이터에만 접근할 수 있으면 되는데요.
NUI 의 SceneView는 2D Content에 들어가는 Component이므로 하나의 Scene에 여러개의 SceneView를 추가할 수 있으며, 여러 SceneView가 하나의 Camera를 돌려 쓸 수 있습니다.
SceneView별로 카메라를 위한 속성이 다르므로, 위에서 말씀드린 종횡비를 포함하여 Capture를 위해 필요한 속성들이 SceneView에 종속적으로 동작하기 때문에, SceneView에서 Capture를 컨트롤 하는 것이 적합합니다.

그 외에도 차치하신 DALi 구조적인 문제로서도 Camera에서 Capture를 제공하기 위해서는 다양한 고통이 필요한데요.
어쨌거나 NUI.Scene3D에서는 대부분의 씬을 위한 기능이 SceneView를 통해 제공되고 있기도 하고, 어쨌거나 해당 SceneView에 포함된 Scene의 렌더링 결과를 얻어내고자 하는 API가 SceneView에 있는 것이 큰 문제가 되지는 않을 것이라고 생각됩니다.

Reply:
>ImageUrl class 는 UI Thread 에서만 생성되어야하는 class 이기 때문에 Task 로 만들 수 없습니다. ImageUrl class 는 외부 요인에 의해 생성된 이미지요소(ex : Texture, NativeImage 등,)의 lifecycle을 관리하기위한 class 이기 때문입니다.

C# Task는 Thread가 아닙니다. C++의 future로 이해해주시면 좋을 것 같습니다.

Reply:
>Camera 가 동시에 여러 SceneView 에 속해있을 수도 **없고**(오타인가요?)
>, 여러 SceneView가 하나의 Camera를 돌려 쓸 수 있습니다.

음.. Camera도 Actor이기 때문에 특정 시점에 하나의 SceneView에 속하는것 아닌가요?

> Camera 자기 자신은 자기가 속해있는 SceneView 도 모르는 상황

``` c++

  /**
   * Retrieve the parent of an Actor.
   * @return The parent actor, or NULL if the Actor does not have a parent.
   */
  Actor* GetParent() const
```
GetParent를 못쓰는건가요?

지금까지 계속 Camera도 Actor이라고 하셔서 당연히 단일 Parent에 속해 있어야 하고, Parent정보에 접근 가능한걸로 생각했습니다.

그러면, Camera는 ViewTree에 속하지 않는 객체인건가요?
이런 설명을 빼먹은체로 계속 Camera도 Actor라고 강조하셔서 아마 저 말고도 많은 분들이 Camera가 ViewTree에 속한다고 생각하시고 있을 것 같네요. 
그러면 Camera는 Actor 성격의 객체가 아닌데 그냥 Actor로 둔건가요?


일단 저는 Camera가 단일 SceneView에만 속하는줄 알았습니다. (타 플렛폼도 그렇지 않나요?)
그래서 Capture를 위한 잡다한 사항을 덮어 두고, 
API를 사용하는 입장에서, Camera는 SceneView에 속하고, 그렇기 때문에 Camera가 자신이 속한 SceneView에 대한 참조를 얻을 수 있다고 생각했습니다. 

그렇게 되면 API가 다른 SceneView에 속한 Camera를 인자로 사용할 상황 자체가 불가능해지기 때문에 더 사용에 실수가 적어지는 API가 된다고 생각했습니다.

Reply:
Camera 도 Actor 이고 ViewTree 에 속해있어야지 정상동작을 합니다.

> 여러 SceneView가 하나의 Camera를 돌려 쓸 수 있습니다.

이건 어디서 비롯된 오해인지 모르겠습니다.

> Camera 자기 자신은 자기가 속해있는 SceneView 도 모르는 상황

Camera 는 dali-core 에 정의된 Actor 이고, SceneView 는 dali-scene3d 에 정의된 CustomActor 이기 때문에 기본적으로 Camera 는 SceneView 의 존재를 모릅니다.

Camera 를 SceneView 의 child 로 추가하는 것은 Scene3D library 를 개발 및 이용하는 사람의 자유이지만, Camera가 *반드시* SceneView 아래에 있다는 것을 보장하지 못합니다. 왜냐하면 Camera 도 Actor 이기 때문에, 다른 Actor 의 child 로 자유롭게 추가가 가능하기 때문입니다.

따라서 자신이 속해있는 SceneView를 알아내기 위해서는 C# 이나 dali-csharp-binder 레벨에서 일일히 자신의 Parent 를 따라 올라가서 Actor 를 Dali::Scene3D::SceneView 로 (내부적으로 dynamic_cast 를 사용하는) DownCast 해나가면서 SceneView 를 알아내야 하는데, 이렇게 Parent 를 따라 올라가는 동작을 추가하는 것은 성능적으로 크나큰 손해이기 때문에 넣고싶지 않습니다.

이런 의미에서 Camera 가 SceneView 를 모른다는 표현을 사용했습니다.

Reply:
Task를 쓰는 SceneView.Capture로 진행하도록 하겠습니다.
+
Capture Finished전에 SceneView SceneOff하는 부분에 대해서 Fail 올리는 것으로 하겠습니다.

Reply:
> C# Task는 Thread가 아닙니다

라고 말씀하셨는데, Task 자체는 worker thread 에서 동작하는 것으로 알고 있습니다.

간단하게 검색한 결과 : https://learn.microsoft.com/en-us/dotnet/api/system.threading.thread.currentthread?view=net-8.0

이런 상황에서, Task 의 return 값으로 , worker thread 에서 생성이 불가능한 ImageUrl 을 return 하는 것이 가능한지 의문입니다.

만약의 여러가지 흑마법을 사용해서 ImageUrl 을 어찌저찌 생성했다고 해도, Capture 의 결과물로 ImageUrl 이 필요한 상황은 현재 뿐으로, 미래에는 어떤 EventArgument 가 추가로 필요할지 모르는 상황에서 API를 섣부르게 Task 형태로 만드는 것은 장기적으로 좋지 않아보입니다.

Reply:
Task (C#에서는 Task가 thread pool의미도 갖고 있고, Future / Promise역할도 합니다. 저기서 언급한 것은 Future/Promise의 역할로 언급한거에요)

아래 지금 API 구조 그대로 해서 Task를 리턴하는 버전의 코드를 만들어봤습니다. 
이 API동작 안하면 안넣으시면 됩니다. 
``` c#
        public Task<ImageUrl> CaptureAsync(Scene3D.Camera camera, Vector2 size)
        {
            TaskCompletionSource<ImageUrl> ret = new TaskCompletionSource<ImageUrl>();
            int captureId = 0;
            void CaptureHandler(object _, CaptureFinishedEventArgs e)
            {
                if (e.CaptureId == captureId)
                {
                    try
                    {
                        if (e.Success)
                        {
                            ret.SetResult(e.CapturedImageUrl);
                        }
                        else
                        {
                            ret.SetException(new InvalidOperationException("Fail to capture"));
                        }
                    }
                    finally
                    {
                        CaptureFinished -= CaptureHandler;
                    }
                }
            };
            captureId = Interop.SceneView.Capture(SwigCPtr, camera.SwigCPtr, Vector2.getCPtr(size));
            return ret.Task;
        }
```

Reply:
> Task 자체는 worker thread 에서 동작하는 것으로 알고 있습니다.

은기님이 언급한 것은 `Task.Run`두고 말하는 것 같네요

Reply:
마침(?) task await async에 관한 제가 작성했던 글이 있어서 공유합니다.
https://myroot.github.io/task-await-and-async/

Reply:
아래는 제가 실제로 비슷한 샘플 및 함수를 만들어본 코드입니다.

https://github.com/hinohie/TizenFX/commit/cbe1d1070335c1c62ce7ce3ba8fde3c8f9ff9060

위 샘플앱을 실제로 구동시키면 앱의 main thread 가 그대로 멈춰버리면서, 재부팅을 하지 않으면 앱이 종료조차 되지 않습니다.

```cs
        public void Activate()
        {
         control.TouchEvent += OnTouch;
        }

        private bool OnTouch(object s, View.TouchEventArgs e)
        {
            if (e.Touch.GetState(0) == PointStateType.Down)
            {
                Tizen.Log.Error("NUI",$"Before await\n");
                int ret = DummyAsync().Result;
                Tizen.Log.Error("NUI",$"End await {ret}\n");
            }
            return true;
        }

        private async Task<int> DummyAsync()
        {
            TaskCompletionSource<int> ret = new TaskCompletionSource<int>();
            EventHandler handler = (s, e) =>
            {
                Tizen.Log.Error("NUI",$"OnRelayout\n");
                ret.SetResult(12);
            };

            control.Relayout += handler;
            control.SizeHeight = control.SizeHeight + 10.0f;
            try
            {
                Tizen.Log.Error("NUI",$"Before await\n");
                return await ret.Task;
            }
            finally
            {
                control.Relayout -= handler;
            }
        }

```

유언은 다음과 같습니다.

```
ERROR: NUI: ViewEvent.cs: OnTouch(927) > Before touch
ERROR: NUI: ControlSample.cs: OnTouch(55) > Before await
ERROR: NUI: ControlSample.cs: DummyAsync(75) > Before await
```

제가 계속해서 main thread 라는 용어를 사용해서 혼선이 생긴 것 같은데.. 실제로 걱정하는 것을 좀 더 정확하게 표현하자면 ecore main loop, 또는 Dali 의 core event process 루프의 block 입니다.

위 코드를 보시면 Touch 이벤트가 끝나기 전 까지 Task 의 결과값을 기다리게 되는데, Task 의 결과를 넣어주는 것은 아무튼 Dali 의 Relayout process 에 의해서 발생합니다. OnTouch 이벤트가 끝나지 않았으니 그 뒤에 수행되어야할 Relayout 이나 UpdateMessage  등도 전달이 안되고 계속 막혀있는 모습을 볼 수 있습니다.

그리고 아쉽게도 현재 달리는 ProcessCoreEvent 가 도는 동안 ProcessCoreEvent 를 한번 더 요쳥한 경우, 그 즉시 loop 를 수행하는 것이 아니라 Idle time 에 호출하도록 함수호출을 순연시킵니다.

즉, 만약 앱에서 CaptureAsync API 를 Native 에서 올라온 이벤트들 (ex : TouchEvent, KeyEvent, ResourceLoaded, Application.OnCreate, Timer.Tick 등) 에서 호출하게 된다면 main thread 가 멈춤을 의미합니다. 

CaptureAsync 기능은 NUI 이벤트콜백에서 사용할 수 없다는 제약사항을 둘 것이 아니면, 제시한대로 구현하면 안될 것으로 보입니다.

Task 기능이 매우 자유롭고 편한 것은 맞지만 그것은 말씀해주신대로 .Net 으로 컴파일 되는 C# 코드에서의 이야기이지, Native 에서는 별로 큰 도움이 안되는 것으로 보입니다.

Reply:
CaptureFinsihed signal 의 경우에는 UpdateRender Thread 에서 작업을 다 마치고 EventThread 로 요청을 보내는 이벤트들 중 하나로 추상화가 되어있는 상황입니다. (ex : Animation.Finished, PropertyNotification.Notify) 이 추상화 레벨을 다 무시해버리고, CaptureFinished 콜백 하나만을 위해서 eventfd 를 새로 뚫을 수는 없습니다.

애초에 위 추상화된 이벤트는 dali-core 에 구현된 내용들이고, 이를 eventfd 등 thread 관련 행정처리(?)를 할 수 있는 dali-adaptor 까지 끌고 내려가는 것은 고된 업무이며, 이로 인해서 얻는 장점은 별로 크지 않아 보입니다.

Reply:
`   int ret = DummyAsync().Result;`
이걸 쓰셨으니까 그렇습니다. 저 속성은 동기적으로 (thread를 블럭하여) 결과를 기다립니다. 

제대로된 사용법으로 사용하면 블럭이 안됩니다.

Reply:
await를 통한 대기는 thread 를 블럭시키지 않습니다. 

<img src=https://i.ytimg.com/vi/GIeTWr2hUSM/maxresdefault.jpg>

Reply:
가이드 주신대로 여러가지 시도를 해본 결과 OnTouch 는 정상적으로 끝나면서 main thread 에서 발생한 이벤트도 정상적으로 받는 것을 확인했습니다.

```cs
        private bool OnTouch(object s, View.TouchEventArgs e)
        {
            if (e.Touch.GetState(0) == PointStateType.Down)
            {
                Tizen.Log.Error("NUI",$"Before ASDF\n");
                ASDF();
                Tizen.Log.Error("NUI",$"End ASDF\n");
            }
            return true;
        }

        private async void ASDF()
        {
            Tizen.Log.Error("NUI",$"Before await\n");
            var task = DummyAsync();
            int ret = await task;
            Tizen.Log.Error("NUI",$"End await {ret}\n");
        }
```

결국은 async 라는 태그가 달린 별도의 함수를 앱에서 알아서 잘 구현해야만 하고, await 이후 ImageUrl 을 받은 사용자가 무엇을 어떻게 할지는 저희가 아직은 신경 안써도 된다고 이해했습니다.
그렇다면 말씀해주신대로 이후의 작업은 C# 언어 개발자의 숙련도 문제이지, Framework 의 문제가 아닌 것으로 이해되었습니다.

```cs
if (e.Success)
{
    ret.SetResult(e.CapturedImageUrl);
}
```
소소하게 이 부분에 대해서, 우선 e.CapturedImageUrl 은 cMemOwned 가 false 인 상태로 만들어졌기 때문에, 이 상태 그대로 SetResult 에 넣으면 안되고, Registry 에 등록을 한번 해야지 정상적인 사용이 가능할 것으로 보입니다. 승호님께서 구현하실 때 참고바랍니다.

Reply:
그런데 다른 질문이 있는데요,

예를 들어 FileIO 에 걸려있는 아무개 Task 라고 하더라도 유저가 원한다면 main loop 에 block 을 걸어서 FileIO 가 끝날때 까지 강제로 기다리게 하는 방법들이 잘 고려가 되어있을 것으로 예상되는데요...

지금의 Capture 처럼 main loop 에 dependency 가 있는 Task 의 경우에는 앱에서 동기적으로 task complete 결과를 기다리는 코드를 사용한 경우 무조건 dead lock 에 걸리게 될 것인데

이러한 사항이 일반적인 것이어서 따로 주석을 달지 않아도 되는건지, 아니면 함수에서 주석으로 적어놓기만 하면 되는건지 궁금합니다.

Reply:
저희가... event로 뭔가를 제공할때도..
main loop에서 Sleep, 또는 while로 점유한체 대기하면 event가 오지 않을 것인데
이러한 사항을 remarks에 적지 않는것과 동일한게 아닐까요?